src/functions/routers/mod.ts

```diff
@@ -0,0 +1,125 @@
+ * @returns {ActionsRouter}
+ */
+export const BlockActionsRouter = <
+  InputParameters extends ParameterSetDefinition,
+  OutputParameters extends ParameterSetDefinition,
+  RequiredInput extends PossibleParameterKeys<InputParameters>,
+  RequiredOutput extends PossibleParameterKeys<OutputParameters>,
+>(
+  func: SlackFunction<
+    InputParameters,
+    OutputParameters,
+    RequiredInput,
+    RequiredOutput
+  >,
+) => {
+  return new ActionsRouter(func);
+};
+
+class ActionsRouter<
+  InputParameters extends ParameterSetDefinition,
+  OutputParameters extends ParameterSetDefinition,
+  RequiredInput extends PossibleParameterKeys<InputParameters>,
+  RequiredOutput extends PossibleParameterKeys<OutputParameters>,
+> {
+  private func: SlackFunction<
+    InputParameters,
+    OutputParameters,
+    RequiredInput,
+    RequiredOutput
+  >;
+  private routes: Array<
+    Array<string[] | RegExp | ActionHandler<typeof this.func.definition>>
+  >;
+
+  constructor(
+    func: SlackFunction<
+      InputParameters,
+      OutputParameters,
+      RequiredInput,
+      RequiredOutput
+    >,
+  ) {
+    this.func = func;
+    this.routes = [];
+  }
+
+  /**
+   * Add an action handler for something that can matche an action_id.
+   */
+  add(
+    action_id: string | string[] | RegExp,
+    handler: ActionHandler<typeof this.func.definition>,
+  ): ActionsRouter<
+    InputParameters,
+    OutputParameters,
+    RequiredInput,
+    RequiredOutput
+  > {
+    if (typeof action_id === "string") {
+      action_id = [action_id];
+    }
+    this.routes.push([action_id, handler]);
+    return this;
+  }
+
+  /**
+   * Returns a method handling routing of action payloads to the appropriate action handler.
+   * The output of export() should be attached to the `blockActions` export of your function.
+   */
+  export(): ActionHandler<typeof this.func.definition> {
+    return async (context) => {
+      const action = context.action.action_id as string;
+      const handler = this.matchHandler(action);
+      if (handler === null) {
+        // TODO: what do in this case?
+        // perhaps the user typo'ed the action id when registering their handler or defining their block.
+        console.warn(
+          `Received block action payload with action_id=${action} but this app has no action handler defined to handle it!`,
+        );
+        return;
+      }
+      if (handler.constructor.name === "AsyncFunction") {
+        return await handler(context);
+      } else {
+        return handler(context);
+      }
+    };
+  }
+
+  /**
+   * Return the first registered ActionHandler that matches the action ID string provided.
+   */
+  matchHandler(
+    action: string,
+  ): ActionHandler<typeof this.func.definition> | null {
+    for (let i = 0; i < this.routes.length; i++) {
+      const route = this.routes[i];
```

Comment:
ultra nit: Assuming each item in `this.routes` is indeed a tuple, we could destructure to give them meaningful names.

```ts
const [constraint, handler] = route;
```